File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var widthOfBinaryTree = function(root) {
    let maxA = 0n, hash=new Map();
    const recurse = (cur, width, level) =>
    {
        if(cur===null)
        {
            return;
        }
        if(hash.has(level))
        { 
            // console.log(hash);
            let [min, max] = hash.get(level);
            if ((BigInt)(min) > (BigInt) (width))
                min = (BigInt) (width);
            else 
                min = (BigInt) (min); 
            
             if ((BigInt) (max) <  (BigInt) (width))
                max = (BigInt) (width);
            else 
                max = (BigInt) (max); 
    
            hash.set(level, [min, max]);
        }
        else{

            hash.set(level, [(width), (width)]);
        }

        if((BigInt) (hash.get(level)[1]) - (BigInt) (hash.get(level)[0]) +1n > maxA)
            maxA = (BigInt) (hash.get(level)[1]) - (BigInt) (hash.get(level)[0]) + 1n;
        else 
            maxA= (BigInt) (maxA);
        recurse(cur.left, (BigInt)  (width)*2n-1n, level+1);
        recurse(cur.right, (BigInt) (width) *2n, level+1);
    }
    recurse(root,1, 0)
    return maxA;
};

File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    Map<Integer, List<Long>> m = new HashMap<>();
    public int widthOfBinaryTree(TreeNode root) {
        if(root == null)
            return 0;
        long maxWidth = 1;
        inorderUtil(root, 1, 1);
        System.out.println(m);
        for(Integer i: m.keySet()) {
            long currWidth = 1;
            currWidth = m.get(i).get(1) - m.get(i).get(0) + 1;
            maxWidth = Math.max(maxWidth, currWidth);
        }
        return (int) maxWidth;
    }
    
    public void inorderUtil(TreeNode root, long width, int level) {
        if(root != null) {
            if(!m.containsKey(level)) {
                m.put(level, new ArrayList<>());
                m.get(level).add(width);
                m.get(level).add(width);
            } else {
                 long min = Math.min(m.get(level).get(0),width);
                 long max = Math.max(m.get(level).get(1),width);
                 m.get(level).set(0, min);
                 m.get(level).set(1, max);
            }
           
            inorderUtil(root.left, 2* width -1, level+1);
            inorderUtil(root.right, 2* width,level+1);
        }  
    }
}